File: Python/DesignPatterns/Singleton_Design_pattern.py
import threading


class Logger:
    _instance = None
    _lock = threading.Lock()

    def __init__(self, file_path: str):
        print("object Created")
        self.file_path = file_path

    @classmethod
    def get_instance(cls) -> "Logger":
        print("Thread")
        with cls._lock:
            if cls._instance is None:
                cls._instance = Logger("filePath")
        return cls._instance


if __name__ == "__main__":
    logger = Logger.get_instance()

    def task() -> None:
        Logger.get_instance()

    thread1 = threading.Thread(target=task)
    thread2 = threading.Thread(target=task)
    thread3 = threading.Thread(target=task)

    thread1.start()
    thread2.start()
    thread3.start()

    thread1.join()
    thread2.join()
    thread3.join()

File: C++/DesignPatterns/Singleton_Design_pattern.cpp
#include <iostream>
#include <memory>
#include <mutex>
#include <string>
#include <thread>

class Logger {
public:
    static Logger& getInstance() {
        std::cout << "Thread\n";
        std::call_once(initFlag, []() {
            instance.reset(new Logger("filePath"));
        });
        return *instance;
    }

    Logger(const Logger&) = delete;
    Logger& operator=(const Logger&) = delete;

private:
    explicit Logger(const std::string& filePath) : filePath(filePath) {
        std::cout << "object Created\n";
    }

    std::string filePath;
    static std::unique_ptr<Logger> instance;
    static std::once_flag initFlag;
};

std::unique_ptr<Logger> Logger::instance = nullptr;
std::once_flag Logger::initFlag;

int main() {
    Logger& logger = Logger::getInstance();
    (void)logger;

    auto task = []() {
        Logger::getInstance();
    };

    std::thread thread1(task);
    std::thread thread2(task);
    std::thread thread3(task);

    thread1.join();
    thread2.join();
    thread3.join();

    return 0;
}
